crates/proc-macros/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+// Source:
+// https://github.com/starkware-libs/papyrus/tree/main/crates/papyrus_proc_macros
+
+//! This macro is a wrapper around the "rpc" macro supplied by the jsonrpsee library that generates
+//! a server and client traits from a given trait definition. The wrapper gets a version id and
+//! prepend the version id to the trait name and to every method name (note method name refers to
+//! the name the API has for the function not the actual function name). We need this in order to be
+//! able to merge multiple versions of jsonrpc APIs into one server and not have a clash in method
+//! resolution.
+//!
+//! # Example:
+//!
+//! Given this code:
+//! ```rust,ignore
+//! #[versioned_starknet_rpc("V0_7_1")]
+//! pub trait JsonRpc {
+//!     #[method(name = "blockNumber")]
+//!     fn block_number(&self) -> anyhow::Result<u64>;
+//! }
+//! ```
+//!
+//! The macro will generate this code:
+//! ```rust,ignore
+//! #[rpc(server, namespace = "starknet")]
+//! pub trait JsonRpcV0_7_1 {
+//!     #[method(name = "V0_7_1_blockNumber")]
```

Comment:
this does not follow the spec unfortunately, we'd like to have a prefix like `/v0.7.1/` in the URL instead - otherwise our rpcs won't be compatible with the rest of the ecosystem
that's also what juno and pathfinder do iirc

Reply:
Just realized that I'm also not sure if it should be:
* `/rpc/v0.7.0`
* or `/rpc/v0_7_0`

Happy to change - took the dot notation for now since I saw that suggestion on slack.

Reply:
I think this would be cleaner: `http://localhost:9944/rpc/v0_7_0
`

Reply:
Gotcha, updated 🫡